File: test/test_fk_ik.cpp
#include <gtest/gtest.h>
#include <Eigen/Dense>
#include <chrono>
#include <vector>
#include <random>
#include <human_model/human_model.hpp>

#define NUM_TESTS 1e4

class HumanKinematicModelTest : public ::testing::Test {
protected:
  // Member variables
  double shoulder_distance;
  double chest_hip_distance;
  double hip_distance;
  double upper_arm_length;  
  double lower_arm_length;  
  double upper_leg_length;
  double lower_leg_length;  
  double head_distance;     

  size_t n_dof;
  size_t n_param;
  Eigen::VectorXd q;
  Eigen::VectorXd q2;
  Eigen::VectorXd param;
  Eigen::VectorXd param2;
  std::vector<human_model::JointLimits> qbounds;
  human_model::keypoints kp_in_ext;
  human_model::keypoints kp2_in_ext;
  human_model::keypoints diff_in_ext;
  Eigen::Vector4d chest_q_rotated;

  std::random_device rd;  // Seed for the random number engine
  std::mt19937 gen; // Standard mersenne_twister_engine seeded with rd()

  // Constructor
  HumanKinematicModelTest()
    : shoulder_distance(0.3),
      chest_hip_distance(0.4),
      hip_distance(0.25),
      upper_arm_length(0.3),
      lower_arm_length(0.3),
      upper_leg_length(0.35),
      lower_leg_length(0.4),
      head_distance(0.4),
      n_dof(28),
      n_param(8),
      q(n_dof),
      q2(n_dof),
      param(n_param),
      param2(n_param),
      gen(rd()) {}

  // SetUp function
  void SetUp() override {
    // Initialize class members
    q.setZero();
    q2.setZero();
    param << shoulder_distance, chest_hip_distance, hip_distance, upper_arm_length,
             lower_arm_length, upper_leg_length, lower_leg_length, head_distance;
    param2.setZero();
    std::map<std::string, Eigen::Vector3d> init_keypoints({
      {"head", Eigen::Vector3d::Zero()},
      {"left_shoulder", Eigen::Vector3d::Zero()},
      {"left_elbow", Eigen::Vector3d::Zero()},
      {"left_wrist", Eigen::Vector3d::Zero()},
      {"left_hip", Eigen::Vector3d::Zero()},
      {"left_knee", Eigen::Vector3d::Zero()},
      {"left_ankle", Eigen::Vector3d::Zero()},
      {"right_shoulder", Eigen::Vector3d::Zero()},
      {"right_elbow", Eigen::Vector3d::Zero()},
      {"right_wrist", Eigen::Vector3d::Zero()},
      {"right_hip", Eigen::Vector3d::Zero()},
      {"right_knee", Eigen::Vector3d::Zero()},
      {"right_ankle", Eigen::Vector3d::Zero()}
    });
    kp_in_ext.set_keypoints(init_keypoints);
    kp2_in_ext.set_keypoints(init_keypoints);
    diff_in_ext.set_keypoints(init_keypoints);
    chest_q_rotated.setZero();

    // Initialize joint limits
    // (Vector Initialization: The vector qbounds is initialized with n_dof elements,
    // each set to human_model::JointLimits(-M_PI, M_PI) -> 1 full rotation)
    qbounds = std::vector<human_model::JointLimits>(n_dof, human_model::JointLimits(-M_PI, M_PI));
    human_model::Human28DOF::setDefaultJointLimits(qbounds);
  }
};


// Define a test case
TEST_F(HumanKinematicModelTest, TestFkIk) {
  // Start timing
  auto start = std::chrono::high_resolution_clock::now();

  // Test the fk and ik functions
  for (size_t idx=0; idx<NUM_TESTS; idx++)
  {
    std::cout << std::endl << "=======================================";
    std::cout << "=======================================" << std::endl;
    std::cout << "idx: " << idx+1 << std::endl;

    // Set a random configuration
    for (size_t i=0; i<n_dof; i++)
    {
      std::uniform_real_distribution<> dis(qbounds[i].min_, qbounds[i].max_);
      q[i] = dis(gen);
    }

    // Normalize the quaternion
    q.block(3,0,4,1)/=q.block(3,0,4,1).norm();

    // If the scalar part of the quaternion is negative,
    // multiply by -1 to ensure consistent representation
    if (q(6, 0) < 0) {
      q.block(3,0,4,1) *= -1.0;
    }

    std::cout << "q [original]           : \n" << q.transpose() << std::endl;
    std::cout << "\nparam [original]       : \n" << param.transpose() << std::endl;

    // Compute the forward kinematics and inverse kinematics
    human_model::Human28DOF::fk(q,param,kp_in_ext);
    human_model::Human28DOF::ik(kp_in_ext,qbounds,q,q2,param2,chest_q_rotated); // use q as previous configuration

    std::cout << "\nq2 [before fk]           : \n" << q2.transpose() << std::endl;
    std::cout << "\ndiff q [before fk]       : \n" << (q-q2).transpose() << std::endl;
    std::cout << "\nparam2 [before fk]       : \n" << param2.transpose() << std::endl;
    std::cout << "\ndiff param [before fk]   : \n" << (param-param2).norm() << std::endl;

    // Compute again the forward kinematics
    human_model::Human28DOF::fk(q2,param2,kp2_in_ext);

    Eigen::VectorXd diff_q = q - q2;


    std::cout << "\nq  [after fk]  0-9 : \n" << q.head(10).transpose() << std::endl;
    std::cout << "\nq2 [after fk]      : \n" << q2.head(10).transpose() << std::endl;
    std::cout << "\ndiff q             : \n" << diff_q.head(10).transpose() << std::endl;
    
    for (size_t iarm=0;iarm<4;iarm++)
    {
      std::cout << "\nq  [after fk]  arm : \n" << q.block(10+iarm*4,0,4,1).transpose() << std::endl;
      std::cout << "\nq2 [after fk]      : \n" << q2.block(10+iarm*4,0,4,1).transpose() << std::endl;
      std::cout << "\ndiff q             : \n" << diff_q.block(10+iarm*4,0,4,1).transpose() << std::endl;
    }
    std::cout << "\nq  [after fk]  26-27 : \n" << q.tail(2).transpose() << std::endl;
    std::cout << "\nq2 [after fk]        : \n" << q2.tail(2).transpose() << std::endl;
    std::cout << "\ndiff q               : \n" << diff_q.tail(2).transpose() << std::endl;
    
    
    std::cout << "\nparam2 [after fk]       : \n" << param2.transpose() << std::endl;
    std::cout << "\ndiff param  : \n" << (param-param2).transpose() << std::endl;
    
    // Compute the keypoint distance, configuration distance, and param distance
    double kpt_distance = human_model::keypoints::keypointDistance(kp_in_ext,kp2_in_ext,diff_in_ext);
    double q_distance = (q-q2).norm();
    double param_distance = (param-param2).norm();

    std::cout << "\nkp          : \n" << kp_in_ext << std::endl;
    std::cout << "\nkp2         : \n" << kp2_in_ext << std::endl;
    std::cout << "\nkp diff     : \n" << diff_in_ext << std::endl;
    std::cout << "\nkeypoint distance: \n" << kpt_distance << std::endl;
    std::cout << "\nconfiguration distance: \n" << q_distance << std::endl;
    std::cout << "\nparam distance: \n" << param_distance << std::endl << std::endl;

    // Check if the keypoint distance, configuration distance, and param distance are less than the threshold
    EXPECT_LT(kpt_distance, 1e-8) << "keypoint distance is greater than threshold";
    EXPECT_LT(q_distance, 1e-8) << "configuration difference is greater than threshold";
    EXPECT_LT(param_distance, 1e-8) << "param difference is greater than threshold";

    if (kpt_distance >= 1e-8)
    {
      break;
    }
  }

  // End timing
  auto end = std::chrono::high_resolution_clock::now();
  std::chrono::duration<double> elapsed = end - start;
  std::cout << "Total execution time: " << elapsed.count() << " seconds" << std::endl;
}


int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: test/python/test_fk_ik.py
import numpy as np
from human_kinematic_model import Keypoints, HumanProcess

TEST_Q = np.array([ 0.680375, -0.211234,  0.566198,  0.485962,  0.670301,
                   -0.492489, -0.268313,  0.536459, -0.444451,  0.10794,
                   -0.0452059, 0.257742, -0.270431,  0.0268018, 0.904459,
                    0.83239,   0.271423,  0.434594, -0.716795,  0.213938,
                   -0.967399, -0.514226, -0.725537,  0.608354, -0.686642,
                   -0.198111, -0.740419, -0.782382])

TEST_KPTS = {
    "head":           np.array([0.687107, -0.544971  , 0.345802]),
    "left_shoulder":  np.array([0.666024, -0.236367  , 0.419017]),
    "left_elbow":     np.array([0.862801, -0.298988  , 0.636634]),
    "left_wrist":     np.array([0.962695, -0.443399  , 0.879876]),
    "left_hip":       np.array([1.02737 , -0.00312208, 0.599881]),
    "left_knee":      np.array([1.09435 ,  0.168946  , 0.897213]),
    "left_ankle":     np.array([1.09923 ,  0.340047  , 1.25874 ]),
    "right_shoulder": np.array([0.694727, -0.186102  , 0.71338 ]),
    "right_elbow":    np.array([0.948553, -0.0811031 , 0.592767]),
    "right_wrist":    np.array([1.20627 ,  0.0174684 , 0.475016]),
    "right_hip":      np.array([1.00407 , -0.0997844 , 0.829257]),
    "right_knee":     np.array([1.11579 ,  0.225018  , 0.896494]),
    "right_ankle":    np.array([1.12542 ,  0.615157  , 0.80875 ])
}
TEST_KPT = Keypoints()
TEST_KPT.set_keypoints(TEST_KPTS)


def test_kinematics():
    shoulder_distance  = 0.3
    chest_hip_distance = 0.4
    hip_distance       = 0.25

    upper_arm_length = 0.3
    lower_arm_length = 0.3
    upper_leg_length = 0.35
    lower_leg_length = 0.4
    head_distance    = 0.4

    n_dof = 7+3+4*4+2
    n_param = 3+2+2+1

    # DEBUG: set random seed
    np.random.seed(0)

    param = np.array([
        shoulder_distance,
        chest_hip_distance,
        hip_distance,
        upper_arm_length,
        lower_arm_length,
        upper_leg_length,
        lower_leg_length,
        head_distance
    ])

    for _ in range(10**4):
        # Randomly generate configuration vector
        # q = np.random.rand(n_dof)
        # q[3:7] /= np.linalg.norm(q[3:7]) # Normalize chest rotation quaternion
        q = TEST_Q

        # Create human kinematic model
        model = HumanProcess(n_dof=n_dof, n_params=n_param)
        
        # === Test Forward Kinematics ===
        kpts = model.forward_kinematics(q, param)
        kp_in_ext = Keypoints()
        kp_in_ext.set_keypoints(kpts)

        # Print and assert results
        print("\nkp_in_ext [original]:\n", kp_in_ext)
        print("\nkp_in_ext [test]:\n", TEST_KPT)

        assert np.allclose(kp_in_ext.get_keypoints(), TEST_KPT.get_keypoints(), atol=1.e-4), \
            "kp_in_ext is not equal to TEST_KPT. Error in forward kinematics."
        # ===============================

        # === Test Inverse Kinematics ===
        q2, param2 = model.inverse_kinematics(kp_in_ext)

        # Print and assert results
        print("\nq [original]:\n", q)
        print("\nq2 [after ik]:\n", q2)
        print("\nparam [original]:\n", param)
        print("\nparam2 [after ik]:\n", param2)

        assert all(np.abs(q-q2) < 1e-8), "q2 computed by IK is not equal to original q"
        assert all(np.abs(param-param2) < 1e-8), "param2 computed by IK is not equal to original param"
        # === Test Inverse Kinematics ===






        # === Test forward kinematics ===

        kpts2 = model.forward_kinematics(q2, param2)
        kp_in_ext2 = Keypoints()
        kp_in_ext2.set_keypoints(kpts2)

        # Compute distance between keypoints computed by forward kinematics
        distance, diff_in_ext = Keypoints.keypoint_distance(kp_in_ext, kp_in_ext2)

        # Print and assert results
        print("\n\nkp_in_ext [original]: ", kp_in_ext)
        print("\nkp_in_ext2 [after fk]: ", kp_in_ext2)
        print("\ndiff_in_ext: ", diff_in_ext)
        print("\ndistance: ", distance)

        assert distance < 1e-8, "distance between keypoints is greater than threshold after FK"
        assert np.linalg.norm(param-param2) < 1e-8, "param difference is greater than threshold"

        # # Print results
        # print("\n\nq: ", q)
        # print("q2: ", q2)
        # print("diff q: ", q-q2)
        # print("param: ", param)
        # print("param2: ", param2)
        # print("diff param: ", np.linalg.norm(param-param2))
        # print("kp_in_ext: ", kp_in_ext)
        # print("kp_in_ext2: ", kp_in_ext2)
        # print("diff kp_in_ext: ", diff_in_ext)
        # print("distance: ", distance)

        # # Assert results
        # assert distance < 1e-8, "distance is greater than threshold"


def main():
    test_kinematics()


if __name__ == '__main__':
    main()